packages/components/src/custom-select-control-v2/legacy-adapter.tsx

```diff
@@ -0,0 +1,20 @@
+/**
+ * Internal dependencies
+ */
+import _LegacyCustomSelect from './legacy-component';
+import _NewCustomSelect from './default-component';
+import type { CustomSelectProps, LegacyCustomSelectProps } from './types';
+
+export function isLegacy( props: any ): props is LegacyCustomSelectProps {
+	return typeof props.options !== 'undefined';
+}
+
+export function DefaultExport(
+	props: LegacyCustomSelectProps | CustomSelectProps
```

Comment:
```suggestion
	props: WordPressComponentProps<
		LegacyCustomSelectProps | CustomSelectProps,
		'button'
	>
```

It seems like this is the best place to mix in the button event handler types (and everything else that could be passed onto the final `<button>`), rather than defining them explicitly in the `LegacyCustomSelectProps` type (the `onBlur`, `onFocus`, etc).

I imagine you did it explicitly because the legacy version explicitly passes only that subset of event handlers down to the underlying component, but since we're going with full `...restProps` passing in the new implementation I think it's fine to be honest with our types.

Reply:
I believe I had something like that, but I explicitly added them after a review: https://github.com/WordPress/gutenberg/pull/57902#discussion_r1457221361

But curious if we should go back to it or if I could adjust what I had before to improve it.

Reply:
I don't have a strong opinion about the specifics, only that we have to accept `button` prop types at this top-level adapter _for the non-legacy case as well_, in some way or another.

So for example, this cannot be a TS error:

```jsx
<CustomSelect label="foo" onClick={ () => {} }>
	foobar
</CustomSelect>
```

Reply:
Makes sense!

Reply:
I don't think that the two comments (mine and Lena's) are in opposition — I still believe that, in order to support the same API surface of the legacy component, we should _not_ wrap `LegacyProps` in `WordPressComponentProps`.
At the same time, Lena's suggestion is about supporting those props for the non-legacy version

What about

```
	props:
		| LegacyCustomSelectProps
		| WordPressComponentProps< CustomSelectProps, 'button' >
```

?